doc/a_tool_config.md
# 配置文档

每一个项目的根目录都有一个独立的config.json配置文件，其详细配置如下：

* `"projectPath": null` - 【常用】工程目录前缀

* `"cssDir": "css"` - css文件夹名称

* `"imagesDir": "css/i"` - images文件夹名称

* `"jsDir": "js"` - js文件夹名称

* `"htmlDir": "html"` - html文件夹名称

* `"widgetDir": "widget"` - widget文件夹名称

* `"outputDirName": "build"` - 输出文件夹名称

* `"outputCustom": ""` - 自定义输出文件夹，以逗号分隔的字符串

* `"cdn": "//misc.360buyimg.com"` - 【常用】静态cdn域名

* `"serverDir": "misc.360buyimg.com"` - 上传至远端服务器文件夹的名称

* `"previewServerDir": "page.jd.com"` - html文件夹上传至服务器所在的文件夹名称

* `"widgetServerDir": "jdfwidget.jd.com"` - widget服务器所在的文件夹名称

* `"widgetOutputName": "widget"` - 全局widgetOutputName名称

* `"widgetOutputMode": 1` - 编译全局wigetOutputName模式，共三种：1: all widgets|2: white list|3: black list

* `"widgetWhiteList": []` - 指定白名单，在widgetOutputMode=2时，输出这个列表内容到widget.js/widget.css中

* `"widgetBlackList": []` - 指定黑名单，在widgetOutputMode=3时，排除这个列表的widget

* `"widgetNesting": true` - widget嵌套功能开关，默认开启

* `"localServerPort": 80` - 【常用】本地服务器端口

* `"build"`
	* `"jsPlace": "insertBody"` - 调试时js文件位置 insertHead|insertBody

	* `"livereload":true` - 是否开启liveload

	* `"sass":true` - 是否开启sass编译

	* `"less":true` - 是否开启less编译

	* `"csslint": false` - 是否开启csslint

* `"upload"`
	* `"type": "http"` - 默认 ftp scp http
    * `"host": null` - 服务器的域名或者ip
    * `"user": null` - 上传时使用的用户名, ftp、scp需要，http不需要
    * `"password": null` - 规则同上
    * `"port": null` - 传输端口，ftp默认21，scp默认22，http默认3000
    * `"rootPrefix": "/var/www/html/"` - scp时传输的目录前缀，用来确认上传文件最终的地址，一个文件最终的地址会是rootPrefix + serverDir + projectPath + filePath，你可以根据自身server的配置来修改这个值

* `"output"`
	* `"cssImagesUrlReplace": true` - css中图片url加cdn替换

	* `"linkReplace": true` - 给link.href添加cdn前缀，v3.3.0版本新增，之前由`jsUrlReplace`参数控制link.href添加cdn

	* `"jsUrlReplace": false` - js文件的id和dependences是否添加cdn前缀

	* `"jsPlace": "insertBody"` - 编译后js文件位置 insertHead|insertBody

	* `"cssCombo": true` - 【常用】css进行combo

	* `"jsCombo": true` - 【常用】js进行combo todo

	* `"hasBanner": true` - 是否给js文件，css文件添加banner时间戳

	* `"compressJs":true` - 【常用】是否开启压缩js文件

	* `"compressCss":true` - 【常用】是否开启压缩css文件

	* `"compressImage":true` - 【常用】是否开启压缩图片

	* `"cssSprite":true` - 是否开启css sprite功能

	* `"cssSpriteMode": 1` - 0: 将所有css文件中的背景图合并成一张sprite图片，1: 将每一个widget中的背景图分别合并成一张图片

	* `"cssSpriteMargin": 10` - css sprite图片之间的间距

	* `"cssSpriteDirection": vertical` - vertical：垂直合并，horizontal：水平合并

	* `"cssAutoPrefixer": true` - 【常用】是否自动删除过时的浏览器css前缀，如`-webkit-box-orient: vertical;`

	* `"browserslist": ["last 2 version", "> 0.2%", "ie > 7"]` - 【常用】配置autoprefixer的browserslist。

	* `"base64": false` - 是否对图片进行base64编码

	* `"webp":false` - 是否生成对应的webp图片

	* `"excludeFiles": ""` - 【常用】想要直接忽略的文件/文件夹，以逗号分隔的字符串："test,build"

* `"babel"` - 默认只启用基本转义(3.4.10及之前preset-es2015，之后为preset-env)，您也可以在项目根目录下新建一个`.babelrc`文件独立配置
	* `"presets": []`
	* `"plugins": []`




CHANGELOG.md
## changelog

### 3.4.12 / 2018/08/15 19:58:00
* [add] 支持autoprefixer的browserslist配置，灵活输出浏览器前缀。

### 3.4.11  / 2018/08/01 14:43:00
* [opti] babel编译默认支持由preset-es2015升级到preset-env。

### 3.4.10  / 2018/03/13 16:42:00
* [fix] 修复对link进行cdn路径替换时，非css link标签，如`<link rel="shortcut icon" type="image/x-icon" href="/favicon.ico">`也会进行cdn替换的bug。

### 3.4.9  / 2018/01/29 13:10:00
* [fix] 修复css中对url路径拼接的一些bug
* [add] 支持css url路径的单引号引用写法

### 3.4.8  / 2018/01/17 18:40:00
* [fix] 修复github提示的安全风险
* [add] 支持{}对象为seajs的factory，比如`define('id', [], {})`

### 3.4.7  / 2018/01/04 14:10:00
* [fix] 修复widgetOutputName在rebuild时没有监听到js css改动的bug
* [fix] 固定browser-sync版本，由于browser-sync在2.23.2版本中，对express的中间件行为有破坏性改动，因此固定bs版本为2.21.0

### 3.4.6  / 2017/12/27 14:39:00
* [fix] 修复压缩时某些配置项失效的bug，比如hasBanner取消时间戳配置。
* [fix] 修复指定目录或文件输出时，widget没有进行编译的bug。
* [opti] 优化`jdf b` rebuild的编译逻辑，减少不必要的编译过程，提高编译速度。具体为：增删改js,css只会编译同一类文件，增删改图片不编译，修改vm或html只编译widget不编译js,css

### 3.4.5  / 2017/12/20 17:15:00
* [opti] 更改node-sass的dependency为`latest`,保证从node-sass最新版本中下载对应node版本的binding文件，避免本地编译(本地编译需额外安装Python,麻烦)
* [fix] 修复jdf b时，如果seajs调用了`seajs-combo`插件而请求combo文件时，define没有提供id的bug。现在jdf b强制取消seajs对js的combo，jdf o/u等命令不受影响。
* [opti] 提供更友好的js错误提示，现在js错误能出到具体文件。

### 3.4.3 / 2017/11/08 17:19:00
* [fix]修复安装时，不支持nodejs@4.0版本的问题

### 3.4.2 / 2017/11/02 16:16:00
* [bug]修复设置全局参数之后，引起命令行状态下执行错误的问题

### 3.4.1 / 2017/11/02 15:53:00
* [add]添加全局参数配置方法

### 3.4.0 / 2017/11/02 13:58:00
* [add]添加`jdf install`命令，可将jdf ui组件安装到本地项目中
* [add]给`jdf install`命令添加`--list`参数，用来显示所有的组件列表
* [add]可以使用api调用jdf的命令

### 3.3.1 / 2017/10/11 14:49:00
* [bug]修复上传初始化的项目中没有任何文件时，会报错的问题
* [bug]修复使用`jdf u -P`命令时，html引用资源路径不会combo的问题
* [add]支持压缩es6语法

### 3.3.0 / 2017/09/30 18:16:00
* [change] cdn替换逻辑梳理明确，增加config.output.linkReplace参数。在此之前config.output.jsUrlReplace参数承担了三个工作：js文件内cdn替换，script标签cdn替换，link标签cdn替换，考虑到js的相互依赖，js文件内cdn替换和script标签cdn替换依旧共用jsUrlReplace，独立出linkReplace参数，作为link标签替换cdn的选项。
* [fix] 优化在没有widget文件夹时，`jdf b/o`时命令行的提示，避免引起误解。

### 3.2.19 / 2017/09/22 18:19:00
* [fix] 修复jdf o -d/-p时参数失效的bug
* [bug]修复执行jdf u命令时，无法上传合并的widget css文件的问题
* [bug]修复某些情况下安装jdf时会抛异常的问题
* [opti] js解析报错时，提供明显的定位，并强制退出。取消以前版本在verbose模式下打印js错误。

### 3.2.16 / 2017/09/08 16:45:00
* [fix] 现在编译.babel后缀文件不需要在工程中安装babel-preset-es2015了，如果在config.json里主动声明，那还是需要在工程中安装。

### 3.2.11 / 2017/08/29 17:24:00
* [fix] 【紧急】修复在不存在vm的情况下，导致页面不编译的bug
* [change] jdf u默认将不上传html，jdf u -P维持原样
* [add] 重磅！新增widget嵌套功能，widget组合更灵活
* [add] 增加widget preview功能
* [fix] 识别data里json数据，现在data可以支持任意合法的json串

### 3.2.7 / 2017/08/11 17:46:00
* [fix] 插件系统调用jdf工程下的node_modules
* [add] jdf内部增加运行时标记

### 3.2.6 / 2017/08/10 18:27:00
* [fix] 修复在删除.git,.idea等非工程文件里的文件时，自动编译的bug
* [fix] 修复在页面引用两个写法一模一样的widget时，只输出一个widget内容的bug

### 3.2.5 / 2017/08/04 17:08:00
* [add] 增加插件模块说明文档
* [add] 增加钩子，“编译后”
* [add] 增加插件模块，保留了三个钩子，分别是“编译前”，“widget模板渲染前”，“widget模板插入html前”

### 3.2.4 / 2017/07/31 17:08:00
* [add] 创建widget时，提供创建smarty模板选项，使用方法jdf widget --create widgetName --smarty 或 jdf w -sc widgetName
* [add] 支持widget多行定义，widget内的属性现在可以分写在多行

### 3.2.3 / 2017/06/29 17:38:00
* [bug]修复利用widgetOutputName将widget合并之后，html引用合并的文件路径问题

### 3.2.1 / 2017/06/12 10:25:00
* [bug]修复widget合并之后，html引用合并的文件路径问题

### 3.1.7 / 2017/05/05 16:11:00
* [fix]修复jdf b时，删除工程文件时会报错退出的问题
* [add]增加jdf init -c/--current命令，表示在当前文件夹下建立jdf工程
* [add]新建widget时，vm,js等文件默认确认为"y"，回车即可创建

### 3.1.5 / 2017/04/18 15:27:00
* [fix]执行jdf b时，实时编译更新widget合并的新文件

### 3.1.4 / 2017/04/17 15:59:00
* [bug]修复使用seajs.use加载文件时，文件路径处理错误的问题

### 3.1.0 / 2017/04/07 09:38:00
* [fix]修改require('a.js')，以字母开头的路径直接添加cdn路径
* [add]默认添加babel-preset-es2015依赖

### 3.0.10 / 2017/03/15 16:26:00
* [bug]修复无法替换带有后缀的css backgorund url

### 3.0.9 / 2017/01/22 18:41:00
* [fix] 修复压缩png文件过多时电脑过卡的bug
* [add] 提供开启多线程压缩的配置
* [fix] jdf b/o/u/w 不再报错，提供友好提示信息
* 修复输出时，cssSprite遇到不存在的文件会抛异常的bug
* 升级jdf图片压缩工具集
* [fix] 在config.json中配置localServerPort，指定端口运行jdf b
* [add] 实现全局widgetOutputName，并增加说明文档

### 3.0.8 / 2017/01/05 17:39:00
* 扩展velocity #parse，若.vm, .tpl文件是属于某个widget，那么也会将这个widget的js，css文件引用进来

### 3.0.7 / 2017/01/05 12:48:00
* [fix] seajs define无require无依赖时添加id
* [fix] 扩展velocity #parse功能，现在可以正确parse后缀为vm和tpl的velocity文件

### 3.0.6 / 2017/01/03 16:44:00
* 当cdn配置为static.360buyimg.com时，require的jdf组件cdn还应该是misc.360buyimg.com

### 3.0.3 / 2016/12/28 12:51:00
* [add]urlReplace.js集成jsAst

### 3.0.3 / 2016/12/22 11:04:00
* [add] 增加{%widgetOutputName="" type="" %}标签，可以指定本页面js，css的输出文件名
* [fix] 修复seajs依赖匹配错误
* [fix] 修复本地服务器解析combo路径的bug
* [add] jdf b 支持编译babel（需要在项目目录安装babel相关包）

### 3.0.1 / 2016/12/22 9:10:00
* [add]重新设计的底层架构，力求所有操作都在内存中进行，提升速度至少三倍以上
* [add]执行每一条命令，默认都会输出每一个关键点的编译过程信息
* [add]更漂亮的文件浏览列表，手机扫描二维码即可直接打开当前目录
* [add]本地编辑代码，自动刷新已经打开当前页面的所有设备，无需再安装任何额外插件
* [add]在任意一台设备上的任何操作，都会自动同步到其它所有设备上，包括：屏幕滚动，目录切换，页面点击等
* [fix]jdf现在可以精确的输出每一个文件的操作信息
* [add]输出或者上传时支持指定文件通配符
* [add]压缩文件时，jdf会自动根据当前文件数量来决定是否启动多线程
* [add]支持各种图片格式：jpg，png，gif，webp等
* [add]上传时支持多种模式：http，ftp，sftp

### 2.0.10 / 2016-11-14 18:00:00
* [add] 增加.editorconfig的支持
* [add] widget模块集成jdf-log
* [add] 删除jdf release相关命令以及入口
* [add] 删除原来的log.js模块以及在入口文件的引用
* [add] 集成jdf-log模块
* [add] 将所有的file.js和base.js替换成jdf-file组件
* [fix] ftp上传时timeout的错误
* [add] 分离上传模块，支持ftp scp http三种方式，通过jdf u -h查看
* [add] widget模块的install,list,publish重构

### 2.0.9 / 2016-11-04 11:11:11
* [add]widget的type可根据逗号","分隔符生成相应的类型，例：type="vm, js"
* [fix]开启html注释删除时误删IE条件注释

### 2.0.9 / 2016-11-4 11:19:00
* [bug]修复处理require资源时路径错误
* [bug]修复处理seajs.use时的路径错误

### 2.0.9 / 2016-11-3 18:25:00
* [bug]修复css文件中背景图片路径的处理问题
* [bug]修复转换base64编码时抛异常的问题

### 2.0.7 / 2016-9-14 10:50:00
* [add]支持图片生成webp格式，并更新相关css图片链接

### 2.0.3 / 2016-7-6 13:50:00
* [bug]修复无法输出类似icon-null.png的文件名称
* [add]支持输出font icon

### 2.0.1 / 2016-4-28 16:39:00
* [add]cssSprite支持水平合并
* [add]可以指定需要合并的widget名称

### 2.0.0 / 2016-4-20 14:26:00
* 发布jdf@2.0.0

### 1.8.42 / 2016-4-11 16:35:00
* [add]jdf build support combo url
* [add]support the CMD
* [add]cssSprite支持rem

### 1.8.4 / 2016-1-20 14:57:00
* [add]支持nodejs5.0
* [fix]基于一个标准jdf工程，重新整理单元测试

lib/config.js
"use strict";
const path = require('path');
/**
 * @统一配置文件
 */

/**
 * @config Json File Content
 */

module.exports = {
	"compressThreadCrisis": 200, //当要压缩的文件超过这个值时，会开启多进程压缩。

	"configFileName": "config.json", //配置文件名称

	"projectPath": "", //工程目录前缀

	"cssDir": "css", //css文件夹名称
	"imagesDir": "css/i", //images文件夹名称
	"jsDir": "js", //js文件夹名称
	"htmlDir": "html", //html文件夹名称
	"widgetDir": "widget", //widget文件夹名称

	"outputDirName": "build", //输出的目标文件夹名称
	"outputCustom": "", //输出的源文件夹名称，以逗号分隔

    "widgetOutputName": "widget",
    "widgetOutputMode": 1, // 1: all widgets|2: white list|3: black list
    "widgetWhiteList": [], // 指定白名单，在widgetOutputMode=2时，输出这个列表内容到widget.js/widget.css中
	"widgetBlackList": [], // 指定黑名单，在widgetOutputMode=3时，排除这个列表的widget
	"widgetNesting": true, // widget嵌套功能
	"widgetLabel": true, // 给widget嵌入html时打上标签

	"localServerPort": 80, //本地服务器端口

	"cdn": "//misc.360buyimg.com", //静态cdn域名

	"serverDir": "misc.360buyimg.com", //上传至远端服务器文件夹的名称
	"previewServerDir": "page.jd.com", //html文件夹上传至服务器所在的文件夹名称
	"widgetServerDir": "jdfwidget.jd.com", //widget服务器所在的文件夹名称

	"build":{
		"jsPlace": "insertBody", //调试时js文件位置 insertHead|insertBody
		"livereload":true, //是否开启browser-sync监测文件变化
		"sass":true,//是否开启sass编译
		"less":true,//是否开启less编译
		"csslint":false//是否开启csslint
	},
    "upload": {
	    "type": "http",     // 默认 ftp scp http
        "host": null,       // 服务器的域名或者ip
        "user": null,   // 上传时使用的用户名, ftp、scp需要，http不需要
        "password": null,   // 规则同上
        "port": null,       // 传输端口，ftp默认21，scp默认22，http默认3000
        "rootPrefix": "/var/www/html/"  // scp时传输的目录前缀，用来确认上传文件最终的地址，一个文件最终的地址会是rootPrefix + serverDir + projectPath + filePath，你可以根据自身server的配置来修改这个值
    },
	"output":{
		"linkReplace": true, // 给link.href添加cdn前缀，它和cssImagesUrlReplace相互独立
		"cssImagesUrlReplace": true,//css中图片url加cdn替换
		"jsUrlReplace": true, //js文件的id和dependences是否添加cdn前缀，添加script.src cdn前缀
		"comboItemCount":2, //在同一个文件夹中，如果 js 或 css 文件数多于此数字，则会 combo
		"cssCombo": true, //css进行combo
		"jsCombo": true, //js进行combo

		"hasBanner": true, //是否给js文件，css文件添加banner时间戳
		"compresshtml": false,//是否开启压缩html文件
		"compressJs": true,//是否开启压缩js文件
		"compressCss": true,//是否开启压缩css文件
		"compressImage": true,//是否开启压缩图片

		"cssSprite": true, //是否开启css sprite功能
		"cssSpriteMode": 1, //0: 将所有css文件中的背景图合并成一张sprite图片，1: 将每一个widget中的背景图分别合并成一张图片
		"cssSpriteMargin": 10, //css sprite图片之间的间距
		"cssSpriteDirection": 'vertical', //vertical：垂直合并，horizontal：水平合并
		"cssAutoPrefixer": true, //是否自动删除过时的浏览器css前缀
		"browserslist": ['last 2 version', '> 0.2%', 'ie > 7'], // autoprefixer的浏览器筛选

		"base64": true, //是否对图片进行base64编码
		"webp":false, //是否生成对应的webp图片

		"excludeFiles": "",//想要直接忽略的文件/文件夹，路径相对于当前项目根目录，以逗号分隔，例如："test,test.css"
	},

	"babel": {
		// 默认只启用基本转义 http://babeljs.io/docs/plugins/preset-env/
		"defaultPresets": [path.resolve(__dirname, "../node_modules/babel-preset-env")],
    "presets": [],
		"defaultPlugins": [],
    "plugins": []
	},

	// 插件  "jdf-cms"
	"plugins": [],

	//依赖的组件列表
	"components": {}
}
